asllib/doc/ASLTypingReference.tex

```diff
@@ -433,6 +471,154 @@ \section{Static Environments}
 
 This is related to \identd{JRXM} and \identi{ZTMQ}.
 
+\section{Constrained Types}
+\begin{itemize}
+  \item A \emph{constrained type} is a type whose definition depends on an expression, for example, certain integers and bitvectors.
+  \item A type which is not constrained is \emph{unconstrained}.
+  \item A constrained type with a non-empty constraint is \emph{well-constrained}.
+  \item An \emph{under-constrained integer type} is an implicit type of a subprogram parameter.
+  \end{itemize}
+The widths of bitvector storage elements are constrained integers.
+
+We define the following helper predicates to classify integer types:
+\[
+  \isunconstrainedinteger(\overname{\staticenvs}{\tenv} \aslsep \overname{\ty}{\vt}) \aslto \Bool
+\]
+
+We define the following shorthands for classifying integers:
+\[
+  \begin{array}{rcl}
+  \isunconstrainedinteger(\tenv, \vt) &\triangleq& \vt = \TInt(\unconstrained)\\
+  \isunderconstrainedinteger(\tenv, \vt) &\triangleq& \vt = \TInt(\underconstrained)\\
+  \iswellconstrainedinteger(\tenv, \vt) &\triangleq& \vt = \TInt(c) \land \astlabel(c)=\wellconstrained\\
+\end{array}
+\]
+
+\subsection{Comments}
+    This is related to \identd{ZTPP}, \identr{WJYH}, \identr{HJPN}, \identr{CZTX}, \identr{TPHR}
+
+\section{ASL Type System}
+\hypertarget{def-annotaterel}{}
+The type system of ASL is given by the relation $\annotaterel$, which is defined as the disjoint union
+of the partial functions and relations defined in this document.
+The partial functions and relations in this document are defined, in turn, via type system rules.
+
+The output configurations of type assertions have two flavors:
+\begin{description}
+  \item[Normal Outputs.] \hypertarget{def-normal-type-outputs}{}
+  Configurations are typically tuples with different combinations
+  of \emph{static environments}, types, and Boolean values.
+
+  \hypertarget{def-typeerror}{}
+  \item[Type Errors.] Configurations in $\TypeError(\texttt{<string>})$
+  represent type errors, for example, using an integer type as a condition expression, as in \verb|if 5 then 1 else 2|.
+  The ASL type system is set up such that when these \emph{type error configurations} appear,
+  the typing of the entire specification terminates by outputting them.
+\end{description}
+
+We define the mathematical type of type error configurations as follows:
```

Comment:
I think it might be useful to have a small paragraph explaining why we need a mathematical type for type errors - otherwise this might read as a little too abstract?